### PR TITLE
Upgrade HAMT and AMT diffing to more efficient implementation when able to

### PR DESCRIPTION
**Pull request by frrist**
This PR replaces the slow generic ADT diffing with optimized diffing for v3 HAMTs and AMTs when permitted, else it falls back to the generic diffing implementation.

Examples of when we call back to the generic diffing implementation include:
- diffing HAMTs with different bitwidths or hash functions
- diffing when any HAMT is from go-hamt-ipld/v2, go-hamt-iplds/v3 implementation cannot unmarshal the v2 implementation pointer type.
- diffing AMTs with different bitwidths
- diffing when any amt is from go-amt-ipld/v2, go-amt-iplds/v3 implementation cannot unmarshal the v2 implementation pointer type. 

Recomend looking at each commit individually while reviewing.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -46,6 +46,7 @@ require (
 	github.com/lib/pq v1.9.0
 	github.com/libp2p/go-libp2p-core v0.8.5
 	github.com/minio/blake2b-simd v0.0.0-20160723061019-3f5f724cb5b1
+	github.com/minio/sha256-simd v1.0.0
```
Comment (frrist) [inline go.mod:49]: Required due to v0 of specs-actors hamt https://github.com/filecoin-project/sentinel-visor/pull/519/commits/cf343479516df95cf1360cf7f99d1d19a87d70fc#diff-49836f46c8b64b9409ed3f44f1129c1b6bff9777bfc94f9b3d01bcf00ee5e4f8R131

File: vector/runner.go [hunk 2]
```diff
@@ -566,6 +567,12 @@ func modelTypeFromTable(tableName string, expected json.RawMessage, actual []int
 			}
 			actType = append(actType, act)
 		}
+		sort.Slice(actType, func(i, j int) bool {
```
Comment (frrist) [inline vector/runner.go:594]: Ensures we compare results in the correct order while running test vectors. The new H/AMT diffing logic returns elements in a different order than the generic diffing implementation.

**Comment by iand** [comment]
What diffing is used when using the current lotus api lens?

**Comment by frrist** [comment]
The optimized versions when possible, else the generic version. The code path for this will be the same regardless of the lens used.

File: chain/actors/adt/diff/generic.go [hunk 3]
```diff
@@ -21,12 +22,12 @@ type AdtArrayDiff interface {
...
 // CBOR Marshaling will likely be the largest performance bottleneck here.
 
-// DiffAdtArray accepts two *adt.Array's and an AdtArrayDiff implementation. It does the following:
+// GenericArray accepts two *adt.Array's and an AdtArrayDiff implementation. It does the following:
 // - All values that exist in preArr and not in curArr are passed to AdtArrayDiff.Remove()
 // - All values that exist in curArr nnd not in prevArr are passed to adtArrayDiff.Add()
 // - All values that exist in preArr and in curArr are passed to AdtArrayDiff.Modify()
 //  - It is the responsibility of AdtArrayDiff.Modify() to determine if the values it was passed have been modified.
-func DiffAdtArray(preArr, curArr Array, out AdtArrayDiff) error {
+func GenericArray(preArr, curArr adt.Array, out AdtArrayDiff) error {
```
Comment (iand) [inline chain/actors/adt/diff/generic.go:30]: Maybe rename to `CompareArray`?

File: chain/actors/adt/diff/generic.go [hunk 4]
```diff
@@ -79,7 +80,7 @@ type AdtMapDiff interface {
 	Remove(key string, val *typegen.Deferred) error
 }
 
-func DiffAdtMap(preMap, curMap Map, out AdtMapDiff) error {
+func GenericMap(preMap, curMap adt.Map, out AdtMapDiff) error {
```
Comment (iand) [inline chain/actors/adt/diff/generic.go:83]: Maybe rename to `CompareMap`?

File: chain/actors/adt/diff/generic.go (hunk 3 above)
Comment (iand) [reply chain/actors/adt/diff/generic.go:30]: Also rename `AdtArrayDiff` to `ArrayComparer` or `ArrayDiffer`?

File: chain/actors/adt/diff/generic.go (hunk 4 above)
Comment (iand) [reply chain/actors/adt/diff/generic.go:83]: Also rename `AdtMapDiff` to `MapComparer` or `MapDiffer`?

File: chain/actors/builtin/miner/diff.go [hunk 5]
```diff
@@ -117,11 +209,69 @@ func (m *sectorDiffer) Modify(key uint64, from, to *cbg.Deferred) error {
...
+func (m *sectorDiffContainer) Remove(key uint64, val *cbg.Deferred) error {
 	si, err := m.pre.decodeSectorOnChainInfo(val)
 	if err != nil {
 		return err
 	}
 	m.Results.Removed = append(m.Results.Removed, si)
 	return nil
 }
+
+func SectorsAmtBitwidth(c cid.Cid) (int, error) {
```
Comment (iand) [inline chain/actors/builtin/miner/diff.go:221]: Can we generate this and similar functions for other actors?

**Review by iand — COMMENTED** [review]
Haven't finished the review, but committing these comments to unblock any changes

File: chain/actors/builtin/miner/diff.go (hunk 5 above)
Comment (frrist) [reply chain/actors/builtin/miner/diff.go:221]: We can if specs-actors remains consistent in their usage of constants for defining different AMT Bitwidths. Would you prefer this method to be generated on the miner shim instead? (And similarly for the market actor?)

Comment (iand) [reply chain/actors/builtin/miner/diff.go:221]: That would be cool I think

File: chain/actors/adt/diff/generic.go (hunk 4 above)
Comment (frrist) [reply chain/actors/adt/diff/generic.go]: done.

File: chain/actors/adt/diff/generic.go (hunk 3 above)
Comment (frrist) [reply chain/actors/adt/diff/generic.go]: done.

File: chain/actors/adt/adt.go [hunk 6]
```diff
@@ -27,3 +37,102 @@ type Array interface {
...
+	// https://github.com/filecoin-project/specs-actors/blob/v0.9.14/actors/util/adt/map.go#L22
+	case builtin0.InitActorCodeID:
+		fallthrough
+	case builtin0.StorageMarketActorCodeID:
+		fallthrough
+	case builtin0.StorageMinerActorCodeID:
+		fallthrough
+	case builtin0.MultisigActorCodeID:
+		fallthrough
+	case builtin0.StoragePowerActorCodeID:
```
[the remark this answers is not in the document]
Comment (frrist) [reply chain/actors/adt/adt.go]: done.

File: chain/actors/builtin/miner/diff.go (hunk 5 above)
Comment (frrist) [reply chain/actors/builtin/miner/diff.go]: Done.

**Review by iand — APPROVED** [review]
Great work! :rocket:

**Pull request merged**